src/main/scala/com/fulcrumgenomics/bam/DownsampleAndNormalizeBam.scala

```diff
@@ -0,0 +1,206 @@
+import com.fulcrumgenomics.cmdline.{ClpGroups, FgBioTool}
+import com.fulcrumgenomics.commons.util.LazyLogging
+import com.fulcrumgenomics.sopt.{arg, clp}
+import com.fulcrumgenomics.util.{Io, ProgressLogger, Sorter}
+import htsjdk.samtools.util.IntervalList.IntervalMergerIterator
+import htsjdk.samtools.util.{CoordMath, Interval, IntervalList, Murmur3, OverlapDetector}
+
+/**
+  * Companion object with helper classes.
+  */
+private object DownsampleAndNormalizeBam extends LazyLogging {
+  /**
+    * Tracks coverage over a set of regions.
+    */
+  class CoverageTracker(regions: Seq[Interval],
+                        val targetCoverage: Int,
+                        val minMapQ: Int
+                       ) {
+
+    // Overlap detector for region => coverage
+    private val detector = new OverlapDetector[RegionCoverage](0, 0)
+    regions.foreach { region =>
+      detector.addLhs(new RegionCoverage(region.getContig, region.getStart, region.getEnd), region)
+    }
+
+    /**
+      * Checks to see if a template has reads that contribute coverage to any bases that are currently
+      * below the coverage threshold.  If so, coverage from all acceptable reads is accumulated and the
+      * method returns true.  Otherwise no coverage is accumulated and the method returns false.
+      *
+      * @param t a Template to accumulate coverage from if it covers any bases that are under the coverage threshold
+      * @return true if the template added coverage, false otherwise
+      */
+    def add(t: Template): Boolean = {
+      val recs = t.allReads
+        .filterNot(_.secondary)
+        .filterNot(_.unmapped)
+        .filterNot(_.duplicate)
+        .filter(_.mapq >= minMapQ)
+        .toSeq
+
+      val addsCoverage = recs.exists { rec =>
+        detector.getOverlaps(rec.asSam).iterator().exists { cov =>
+          Range.inclusive(math.max(rec.start, cov.start), math.min(rec.end, cov.end)).exists { pos =>
+            cov(pos) <= targetCoverage
+          }
+        }
+      }
+
+      if (addsCoverage) {
+        val intervals = {
+          val iter = recs
+            .sortBy(r => (r.refName, r.start, r.end))
+            .iterator
+            .map(r => new Interval(r.refName, r.start, r.end))
+          new IntervalMergerIterator(iter, true, false, false)
+        }
+
+        for (interval <- intervals; cov <- detector.getOverlaps(interval).iterator()) {
+          cov.add(interval.getStart, interval.getEnd)
+        }
+      }
+
+      addsCoverage
+    }
+  }
+
+  /**
+    * Coverage counter for a region - all coordinates in the public API are 1-based closed ended.
+    */
+  class RegionCoverage(val chrom: String, val start: Int, val end: Int) {
+    private val offset = start
+    private val counts = new Array[Int](CoordMath.getLength(start, end))
+
+    /** Returns the coverage at the given genomic position, or -1 if the position is not between start:end. */
+    def apply(i: Int): Int = {
+      if (i < start || i > end) -1
```

Comment:
Consider either returning `0` (zero coverage) or `None` (no coverage) instead of `-1`, which feels very Java-like.

Reply:
I think actually I can just go back to having it fail if the index is out of bounds.